internal/app/machined/pkg/controllers/k8s/apiserver_endpoints.go

```diff
@@ -0,0 +1,185 @@
+	"github.com/siderolabs/talos/pkg/machinery/resources/k8s"
+)
+
+// APIServerEndpointsController creates a list of API server endpoints.
+type APIServerEndpointsController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Name() string {
+	return "cluster.APIServerEndpointsController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineTypeType,
+			ID:        pointer.To(config.MachineTypeID),
+			Kind:      controller.InputWeak,
+		},
+		safe.Input[*cluster.Member](controller.InputWeak),
+		safe.Input[*config.MachineConfig](controller.InputWeak),
+		safe.Input[*cluster.Identity](controller.InputWeak),
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: k8s.APIServerEndpointsType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *APIServerEndpointsController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		if _, ok := channel.RecvWithContext(ctx, r.EventCh()); !ok && ctx.Err() != nil {
+			return nil //nolint:nilerr
+		}
+
+		machineConfig, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine config: %w", err)
+			}
+
+			continue
+		}
+
+		machineType, err := safe.ReaderGetByID[*config.MachineType](ctx, r, config.MachineTypeID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine type: %w", err)
+			}
+
+			continue
+		}
+
+		identity, err := safe.ReaderGetByID[*cluster.Identity](ctx, r, cluster.LocalIdentity)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting local identity: %w", err)
+			}
+
+			continue
+		}
+
+		members, err := safe.ReaderListAll[*cluster.Member](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error listing affiliates: %w", err)
+		}
+
+		var endpoints []k8s.APIServerEndpoint
+
+		ce := machineConfig.Config().Cluster().Endpoint()
+		if ce != nil {
+			endpoints = append(endpoints, k8s.APIServerEndpoint{
+				Host: ce.Hostname(),
+				Port: toPort(ce.Port()),
+			})
+		}
+
+		if machineType.MachineType() == machine.TypeControlPlane {
+			endpoints = append(endpoints, k8s.APIServerEndpoint{
+				Host: "localhost",
+				Port: uint32(machineConfig.Config().Cluster().LocalAPIServerPort()),
+			})
+		}
+
+		for it := safe.IteratorFromList(members); it.Next(); {
+			memberSpec := it.Value().TypedSpec()
+
+			if identity.TypedSpec().NodeID == memberSpec.NodeID {
```

Comment:
`identity` can be `nil` here... not sure what is the fix, but I feel we can safely add ourselves to the list.

Reply:
Both `DiscoveryServiceController` and `KubernetesPushController` get local id from `*cluster.Identity` (and do not allow to continue if it's nil, same logic here) so I don't know what else we can use in `MemberSpec` to identify current node.

The problem with adding ourselves to the list is from load balancing perspective - we have `localhost+port` already in the list, so if we add ourselves agin, we two times more likely to go to ourselves than to others, no?

Reply:
we will talk to ourselves always, as it's anyways in the "local" tier, no matter which address we're using. 

what I mean is that in the controller above iirc `identity` might be `nil` if it's not found.